src/nacl/public.py

```diff
@@ -362,14 +369,22 @@ def decrypt(self, ciphertext, encoder=encoding.RawEncoder):
         :param ciphertext: [:class:`bytes`] The encrypted message to decrypt
         :param encoder: The encoder used to decode the ciphertext.
         :return bytes: The original plaintext
+        :raises TypeError: if this SealedBox was created with a
+            :class:`~nacl.public.PublicKey` rather than a
+            :class:`~nacl.public.PrivateKey`.
         """
         # Decode our ciphertext
         ciphertext = encoder.decode(ciphertext)
 
+        # Type ignore: self._private_key is an Optional[bytes], but
+        # crypto_box_seal_open expects it to be a bytes object. It's safe to
+        # ignore this: if it is None, crypto_box_seal_open will raise a
+        # TypeError. This is enforced by
+        # test_sealed_box_public_key_cannot_decrypt.
```

Comment:
Looks like this is a legitimate error and we could be more explicit about it by checking at the start of the function:
```
if self._private_key is None:
    raise TypeError("Must be initialised with a PrivateKey.")
```
Or similar... I don't think allowing the lower `TypeError` to propogate is very clear what has been done wrong.

Reply:
If we wanted to provide much better type safety though, then this class could be refactored into 2 clases, or a `Generic` class, with this method only existing on `PrivateKey` version.

Reply:
Let's separate initial type hinting from changing PyNaCl's API to work better in a type hinted world. The latter is desirable and we should do it where appropriate, but for the moment we're looking for full type hinting of existing surface area (with whatever drawbacks that surface area has).

I'm okay with either the type check @Dreamsorcerer proposes (although you'll need to add a test for that case) or a simple `assert self._private_key is not None`. The latter isn't actually safe since you can turn off runtime assertions, but that's no worse than the current API and teaches mypy what the expected type is at that point.

Reply:
It would also be possible to change the classmethod to use `cls.__new__(cls)` and remove the allowance for `None` in `__init__`. The risk of that approach is that if other logic is added to `__init__` in the future then the classmethod won't get that for free, but I'm okay with that if we want to do it.

Reply:
> Let's separate initial type hinting from changing PyNaCl's API to work better in a type hinted world.

Agreed: I'm trying to avoid touching the logic and machinery where possible.

> I'm okay with either the type check @Dreamsorcerer proposes (although you'll need to add a test for that case)

I think `test_sealed_box_public_key_cannot_decrypt` covers this already---do you agree?

I'd still like to add a more helpful exception message while I'm here. Library style seem to be to use `ensure(...)`, but I don't think mypy will understand the guarantees it enforces without help from a plugin.

Reply:
I also like the use of `Generic`, but maybe that's best saved for when we come to tackle `public.py` proper.

Reply:
I couldn't resist. It does spot the following error:

```python
SealedBox(PublicKey(b"publickey")).decrypt(b"ciphertext")
```

```
src/nacl/public.py:404: error: Invalid self argument "SealedBox[PublicKey]" to attribute function "decrypt" with type "Callable[[SealedBox[PrivateKey], bytes, Type[_Encoder]], Any]"  [misc]
```

Reply:
> Library style seem to be to use ensure(...), but I don't think mypy will understand the guarantees it enforces without help from a plugin.

Yep, it won't work with mypy. It also seems less readable and longer code than just doing it the standard way, so I'd suggest ditching that function personally.